File: C/Questao2_LeitoresEscritores.c
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h>

sem_t mutex, writeblock;
int data = 0, rc = 0;

void* leitor(void* arg) {
    int id = *((int*)arg);
    while (1) {
        sem_wait(&mutex);
        rc++;
        if (rc == 1)
            sem_wait(&writeblock);
        sem_post(&mutex);

        printf("Leitor %d leu o valor: %d\n", id, data);
        sleep(1);

        sem_wait(&mutex);
        rc--;
        if (rc == 0)
            sem_post(&writeblock);
        sem_post(&mutex);

        sleep(2);
    }
    return NULL;
}

void* escritor(void* arg) {
    int id = *((int*)arg);
    while (1) {
        sem_wait(&writeblock);
        data++;
        printf("Escritor %d escreveu o valor: %d\n", id, data);
        sleep(2);
        sem_post(&writeblock);
        sleep(3);
    }
    return NULL;
}

int main() {
    pthread_t r1, r2, w1;
    int id1 = 1, id2 = 2, id3 = 1;

    sem_init(&mutex, 0, 1);
    sem_init(&writeblock, 0, 1);

    pthread_create(&r1, NULL, leitor, &id1);
    pthread_create(&r2, NULL, leitor, &id2);
    pthread_create(&w1, NULL, escritor, &id3);

    pthread_join(r1, NULL);
    pthread_join(r2, NULL);
    pthread_join(w1, NULL);

    return 0;
}

File: Java/Questao2_LeitoresEscritores.java
import java.util.concurrent.Semaphore;

public class Questao2_LeitoresEscritores {
    static int data = 0, rc = 0;
    static Semaphore mutex = new Semaphore(1);
    static Semaphore writeblock = new Semaphore(1);

    static class Leitor extends Thread {
        int id;
        Leitor(int id) { this.id = id; }

        public void run() {
            try {
                while (true) {
                    mutex.acquire();
                    rc++;
                    if (rc == 1)
                        writeblock.acquire();
                    mutex.release();

                    System.out.println("Leitor " + id + " leu: " + data);
                    Thread.sleep(1000);

                    mutex.acquire();
                    rc--;
                    if (rc == 0)
                        writeblock.release();
                    mutex.release();

                    Thread.sleep(2000);
                }
            } catch (InterruptedException e) {}
        }
    }

    static class Escritor extends Thread {
        int id;
        Escritor(int id) { this.id = id; }

        public void run() {
            try {
                while (true) {
                    writeblock.acquire();
                    data++;
                    System.out.println("Escritor " + id + " escreveu: " + data);
                    Thread.sleep(2000);
                    writeblock.release();
                    Thread.sleep(3000);
                }
            } catch (InterruptedException e) {}
        }
    }

    public static void main(String[] args) {
        new Leitor(1).start();
        new Leitor(2).start();
        new Escritor(1).start();
    }
}
